### PR TITLE
fix reading 3D data

### PR DESCRIPTION
**Pull request by dschwoerer**
I haven't tested whether that produces sensible results, but at least no error is thrown on loading.

### CODE REVIEW

**Review by johnomotani — COMMENTED** [review]
Looks good to me. Would be nice to add tests - if you add an argument like `coord_3d` (defaulting to `False`) to the test-data creation functions here https://github.com/boutproject/xBOUT/blob/8f14cb2f80870f0ebf2deaa25f83f1e639a0b3a2/xbout/tests/test_load.py#L250, then parametrize one or a few of the tests. Maybe this one https://github.com/boutproject/xBOUT/blob/8f14cb2f80870f0ebf2deaa25f83f1e639a0b3a2/xbout/tests/test_load.py#L901 but ideally something with the topology (ixseps*, jyseps*) that you'd typically use for a 3d-coordinates sim, so there might be better choices.

**Comment by dschwoerer** [comment]
Actually, this is not only needed for 3D metrics, but for 2D as well, as `dz` is going to be a `Field2D` (currently only in the 3d branch).

I don't have a good mesh yet ... once I get there, I will probably anyway want to add more FCI functionality ...

**Comment by ZedThree** [comment]
Sorry, I meant to push to a separate branch and make a PR into this one!

This now works for BOUT++v5 without 3D metrics and has one simple test. It doesn't check 3D coordinates yet.

I suspect there's some fixes needed for `apply_geometries`: namely the assumption that y (and now z) are 1D coordinates. I don't think that's true in the general case, even in v4, as having non-orthogonal coordinates means it's possible to have Y vary in X.

File: xbout/tests/test_load.py [hunk 1]
```diff
@@ -474,7 +480,7 @@ def create_bout_ds(
 
     # BOUT_VERSION needed so that we know that number of points in z is MZ, not MZ-1 (as
     # it was in BOUT++ before v4.0
```
Comment (dschwoerer) [inline xbout/tests/test_load.py:482]: maybe extend the comment?

**Comment by dschwoerer** [comment]
Thanks for working on this :+1:

**Comment by ZedThree** [comment]
Ok, parameterised the test further for 3D coords.

Note, the build flag `use_metric_3D` is currently missing from the BOUT++ output, so this currently won't actually work for 3D coords... PR for BOUT++ incoming!

**Comment by ZedThree** [comment]
This is failing on uploading the coverage:

```
    {'detail': ErrorDetail(string='Unable to locate build via Github Actions API. Please upload with the Codecov repository upload token to resolve issue.', code='not_found')}
Error: Codecov failed with the following error: The process '/usr/bin/bash' failed with exit code 1
```

File: xbout/geometries.py [hunk 2]
```diff
@@ -180,19 +180,34 @@ def apply_geometry(ds, geometry_name, *, coordinates=None, grid=None):
...
-                f"{2.*np.pi*updated_ds.metadata['ZMAX'] - z0}): using value from dz"
+
+        # In BOUT++ v5, dz is either a Field2D or Field3D.
+        # We can use it as a 1D coordinate if it's a Field3D, _or_ if nz == 1
+        bout_v5 = updated_ds.metadata["BOUT_VERSION"] >= 5.0
+        use_metric_3d = updated_ds.metadata.get("use_metric_3d", False)
+        can_use_1d_z_coord = (nz == 1) or use_metric_3d
+
+        if can_use_1d_z_coord:
+            z = _1d_coord_from_spacing(updated_ds["dz"], zcoord)
```
Comment (johnomotani) [inline xbout/geometries.py:191]: Will this work for a v4 data set with `nz == 1`? I think we need a way to get `updated_ds.metadata["dz"]` for that case.

File: xbout/geometries.py [hunk 3]
```diff
@@ -180,19 +180,34 @@ def apply_geometry(ds, geometry_name, *, coordinates=None, grid=None):
...
+        else:
+            if bout_v5:
+                dz = updated_ds["dz"][0, 0]
```
Comment (johnomotani) [inline xbout/geometries.py:194]: It would be nice to check that `dz` is actually constant here. I think at the moment it always should be, but you never know what someone will want to do in future...
```suggestion
                if not np.all(updated_ds["dz"].min() == updated_ds["dz"].max()):
                    raise ValueError(
                        f"Spacing is not constant. Cannot create z coordinate"
                    )
                dz = updated_ds["dz"][0, 0]
```

File: xbout/tests/test_load.py [hunk 4]
```diff
@@ -947,16 +988,23 @@ def test_combine_along_xy(self, tmpdir_factory, bout_xyt_example_files):
         )
         expected = concat([line1, line2, line3], dim="y", data_vars="minimal")
         expected = expected.set_coords("t_array").rename(t_array="t")
+        vars_to_drop = METADATA_VARS + _BOUT_PER_PROC_VARIABLES
+        if bout_v5:
+            vars_to_drop.remove("dz")
```
Comment (johnomotani) [inline xbout/tests/test_load.py:993]: Why is this necessary? Shouldn't `actual` contain `"dz"` since it was added at l.659? I'm obviously missing something, because the tests pass at the moment!

**Review by johnomotani — COMMENTED** [review]
Thanks for this @dschwoerer @ZedThree 

One other thing that might be nice, although I'm not sure how much of a faff it is, would be to copy `dz` from `metadata` into the main Dataset (as a 'coordinate') for v4 - probably best place for this is right at the beginning of `apply_geometry()` (in `xbout/geometries.py`). Then we can use `ds["dz"]`/`da["dz"]` wherever we like, and hopefully most places wouldn't have to check whether it is scalar/2d/3d. Anywhere where we currently use `metadata["dz"]` would need changing to just `["dz"]` - I don't remember how many of those places there are.

**Comment by ZedThree** [comment]
> Thanks for this @dschwoerer @ZedThree
> 
> One other thing that might be nice, although I'm not sure how much of a faff it is, would be to copy `dz` from `metadata` into the main Dataset (as a 'coordinate') for v4 - probably best place for this is right at the beginning of `apply_geometry()` (in `xbout/geometries.py`). Then we can use `ds["dz"]`/`da["dz"]` wherever we like, and hopefully most places wouldn't have to check whether it is scalar/2d/3d. Anywhere where we currently use `metadata["dz"]` would need changing to just `["dz"]` - I don't remember how many of those places there are.

Yes, I agree! I don't currently fully grok xarray's structure, so I wasn't sure about doing this. I'll investigate

File: xbout/geometries.py (hunk 2 above)
Comment (ZedThree) [reply xbout/geometries.py:191]: I think it should do, yes

File: xbout/geometries.py (hunk 3 above)
Comment (ZedThree) [reply xbout/geometries.py:194]: Good shout :+1: 

xarray should presumably be able to support coordinates that vary across multiple dimensions, so at some point we should be able to set this "properly". This would of course be much easier if BOUT++ wrote out the proper coordinates and not just the spacings and metric!

File: xbout/tests/test_load.py (hunk 4 above)
Comment (ZedThree) [reply xbout/tests/test_load.py:993]: Er, I forget exactly why, but this was causing me issues. I'll double check to see if this is still necessary

**Comment by johnomotani** [comment]
> Yes, I agree! I don't currently fully grok xarray's structure, so I wasn't sure about doing this. I'll investigate

The use of `metadata` is a peculiarity of `xbout`. It exists because it's useful to attach 'metadata' like `ixseps1`, etc., to each `DataArray` variable in the `Dataset`, because `DataArray`s don't keep a reference to their parent. So if you have a method on a `DataArray`, called like `ds["n"].bout.somemethod()`, then it can use the metadata. The rough-and-ready way to decide what counts as 'metadata' was to put all scalars into a `dict`, which is then attached to both the Dataset and all the member DataArrays as an attribute. It would be nicer to make some distinction between 'scalar variables' (which might for example be written out by a PhysicsModel) and actual metadata (like input parameters to Mesh). `dz` is a particularly odd case, which it makes sense to handle specially, we just never got around to it.

File: xbout/geometries.py (hunk 3 above)
Comment (johnomotani) [reply xbout/geometries.py:194]: `xarray` can indeed support multidimensional coordinates (we load for example `Rxy` and `Zxy` as 2d (or now 3d I guess) `R` and `Z` coordinates).

The wrinkle is that it's useful also to have a 1d coordinate associated with each dimension - what `xarray` calls a 'dimension coordinate'. The reason it's useful to have those dimension coordinates defined is that it means that when you do a selection (for example) you don't lose track of where in a dimension the data is defined. `xarray`'s broadcasting machinery can use these 'dimension coordinates' to do clever things (see e.g. http://xarray.pydata.org/en/stable/user-guide/computation.html?highlight=alignment#automatic-alignment).

Especially with 3d coordinates coming in, it might well make sense to assume our `x`, `y`, `z` coordinates may be multidimensional, and just attach a global index 'coordinate', e.g. (loosely) `ds["xind"] = GlobalXIndex()`, `ds["yind"] = GlobalYIndex()`, `ds["zind"] = GlobalZIndex()`. I've thought about doing this before, and just not had a strong enough reason when the regular `y` and `z` coordinates were good 1d coordinates. `x` is already doing this, because `psi` is in general 2d, so BOUT++'s 'x-coordinate' isn't a 1d coordinate (it can be defined differently between PFR and core, or between inner and outer SOL in a double-null), so in `xbout` `x` is just the global index number, and `psi_poloidal` (if I remember the name right) is a 2d coordinate. That'd be a change for a separate PR though!

File: xbout/tests/test_load.py (hunk 4 above)
Comment (ZedThree) [reply xbout/tests/test_load.py:993]: Ah because these are scalars/metadata to drop, and with BOUT++ v5, `dz` doesn't appear there so if we don't remove it then it only appears on one side of the comparison.

If I can work out how to move `dz` out of the metadata into the "real" DataArray, then both v4 and v5 should be able to handle it correctly without special handling here

**Comment by johnomotani** [comment]
PS In addition to `metadata`, any 'coordinates' are attached to each DataArray (as long as they share a dimension with it, or (I think) if they are a scalar). That's why various things like `dx` and `dy` are defined as 'coordinates', so they can be used in DataArray methods.

File: xbout/geometries.py (hunk 3 above)
Comment (ZedThree) [reply xbout/geometries.py:194]: Yes, definitely should be another PR!

I'm struggling a bit to get my head around what's the correct thing to do here. Is the bit below, checking that `nz * dz == 2pi*(ZMAX - ZMIN)`\* strictly necessary? If not, then I think this whole bit could boil down to:

1. Move `dz` from `updated_ds.metadata` to `updated_ds`
2. Use `_1d_coord_from_spacing`, with the proviso that this should actually add `ZMIN` instead of `0.5` to the coordinate
3. Done?

---
\* The correct solution is probably that BOUT++ should error if `ZMIN, ZMAX, dz` are all set

File: xbout/geometries.py (hunk 2 above)
Comment (johnomotani) [reply xbout/geometries.py]: It did fail for v4 input with `nz==1` - I added a test which shows the error (b94e9bc), and then fixed it (456de1e).

File: xbout/tests/test_load.py (hunk 4 above)
Comment (johnomotani) [reply xbout/tests/test_load.py:997]: I think I know where to move `dz` out of metadata - probably easier if I just make the change, unless you've done it already @ZedThree?

Comment (ZedThree) [reply xbout/tests/test_load.py:997]: Yes please, I have just been trying and failing to do it :smile:

File: xbout/geometries.py (hunk 3 above)
Comment (johnomotani) [reply xbout/geometries.py:199]: I think the `+ 0.5` in `_1d_coord_from_spacing()` should stay - at least the one I'm looking at is taking into account the offset between the first cell edge (where the coordinate 'starts') and cell centre (where coordinate is evaluated). So like you say, use `_1d_coord_from_spacing()`, but then just add `ZMIN`.

The check with `ZMAX` isn't strictly necessary, but I think I added it because I got bitten by an input file which defined both `dz` and `ZMAX`, but inconsistently, so I'd like to keep the check in.

> * The correct solution is probably that BOUT++ should error if ZMIN, ZMAX, dz are all set

Agreed :+1:

File: xbout/tests/test_load.py (hunk 4 above)
Comment (johnomotani) [reply xbout/tests/test_load.py:997]: Sorry, we could probably do with some developer docs, but I doubt I'll ever have time to write them...

**Review by ZedThree — APPROVED** [review]

**Comment by ZedThree** [comment]
I've double-checked this works for the 1D conduction and 2D blob2d models in both BOUT++ master and next

File: xbout/geometries.py (hunk 3 above)
Comment (johnomotani) [reply xbout/geometries.py:194]: If I add the check for a constant `dz`
```suggestion
                if not np.all(updated_ds["dz"].min() == updated_ds["dz"].max()):
                    raise ValueError(
                        f"Spacing is not constant. Cannot create z coordinate"
                    )
                dz = updated_ds["dz"][0, 0]
```
will it cause problems (e.g. 3D `dz` that's 'constant' but only up to machine precision)?

Comment (ZedThree) [reply xbout/geometries.py:194]: Could use `allclose` instead to account for that?

Comment (dschwoerer) [reply xbout/geometries.py:194]: I think `==` is fine - we use the same in BOUT++ (`isUniform`):
https://github.com/boutproject/BOUT-dev/blob/next/include/field.hxx#L369

**Review by johnomotani — APPROVED** [review]
